i#7433:  properly initialize memref_t in test helper methods

brettcoon opened the pull request:
Fixes the problem of "initialized" memrefs containing uninitialized fields.

In memref_gen.h, memref_t objects are initialized like this:

`memref_t memref = {}; 
`
However, since memref_t is a union, this form of aggregate initialization only initializes the first declared union member. If the first member doesn't cover all other union members, it may leave parts of the other members uninitialized.

To fix this, a new first union member is added in the form of a byte array sized to match the largest member struct.  This ensures that `= {}` style aggregate initialization of memref_t objects reliably zero-initializes the entire union.

With this change in place, a seemingly unrelated unit test (config_reader_unit_test.cpp) began failing a brittle equality test on floating point values.  Such FP comparisons are unsafe due to their sensitivity to minor code perturbations, so a fix for this test is included in this PR as well to enable the test suite to pass.

Fixes: #7433

derekbruening commented on clients/drcachesim/tests/memref_gen.h:
```diff
@@ -64,7 +64,8 @@ struct memref_with_IR_t {
 {
-    memref_t memref = {};
+    memref_t memref;
+    memref.instr = {};
```
It seems likely this mistake could happen again: should we leverage C++ and just give default 0 values to the memref_t fields?

brettcoon replied:
I thought about that, but I'm not sure it's so simple.  In general, giving each union member default values could lead to conflicting initial values, which is why by my reading C++ just picks the first element with an initializer or else zero-initializes the first element.  Either way, it's only guaranteed to zero the whole union if the chosen element is also the largest.  We could give it a constructor that does a memset, but that's kind of ugly and might have runtime implications.

derekbruening replied:
Could we give it a constructor that does what you're doing here: so we don't have to rely on every declaration remembering to do what you're doing here.

brettcoon replied:
I understand the goal; I'm saying it's harder than you'd think.  What I'm doing here is initializing the specific union member that is about to be used, but that relies on knowing what's ahead in the code.  So there's really no way for the compiler to do that sort of future-usage-specific initialization automatically.

What we _could_ do is initialize the entire union footprint, which is potentially wasteful if we're only intending to use a small part of it (e.g. I'm guessing markers have a much smaller footprint than instructions), but at least it's safe.  As far as I can tell C++ still doesn't make that easy, but I believe there are ways to do it with various levels of ugliness.  My current best idea is to define a local union of all structs in memref_t called (for example) memref_noinit_t, then in the real memref_t definition insert a new very first union element that is both the largest element and the only one with a default initializer, e.g. something like `uint8_t byte_array[sizeof(memref_noinit_t)] = {};`.  I _think_ that will accomplish the goal of making `memref_t m = {};` work robustly, and get past all our various compilers.

derekbruening replied:
> My current best idea is to define a local union of all structs in memref_t called (for example) memref_noinit_t,

"local" to where?
Could you give a code example: not sure what "all structs" means.

Since it's not simple to fix: maybe we should leave it as you have it, where the user needs to initialize by setting the target type field to {}.  But let's at least add a comment to the memref.h decl of the union instructing users to do that, and fix all the wrong types in this file so they're initializing the proper type (else we may have more uninit errors).

brettcoon replied:
After some iteration I have uploaded what I think is a clean solution: instead of creating a duplicate "local" union, I just use the size of memref_instr_t since it's the largest, and the static_assert() makes sure this property is verified.  ASAN passes with this version.

derekbruening commented on clients/drcachesim/tests/memref_gen.h:
```diff
@@ -169,7 +173,8 @@ gen_marker(memref_tid_t tid, trace_marker_type_t type, uintptr_t val)
-    memref_t memref = {};
+    memref_t memref;
+    memref.instr = {};
     memref.instr.type = TRACE_TYPE_THREAD_EXIT;
```
Isn't this the wrong union field: there's a special "exit" field so this should be .exit for all of the accesses in this function.

brettcoon replied:
Done.

derekbruening commented on clients/drcachesim/tests/memref_gen.h:
```diff
@@ -75,7 +76,8 @@ gen_addr(memref_tid_t tid, trace_type_t type, addr_t addr, size_t size = 1)
 {
-    memref_t memref = {};
+    memref_t memref;
+    memref.instr = {};
```
Wait: shouldn't all the fields in this function be .data and not .instr?

brettcoon replied:
Done.

derekbruening commented on clients/drcachesim/tests/memref_gen.h:
```diff
@@ -64,7 +64,8 @@ struct memref_with_IR_t {
 {
-    memref_t memref = {};
+    memref_t memref;
+    memref.instr = {};
```
What is this "gen_addr" anyway: it's trying to set a generic type?  But it doesn't dispatch on type and so can't set all the fields of the appropriate union field; and just setting .instr to 0 is not guaranteed to avoid another uninit error.  It seems this routine has to dispatch on type.

See the gen_data and gen_exit comments: they have the wrong types too.

brettcoon replied:
It appears this is only used in tests/invariant_checker_test.cpp to generate TRACE_TYPE_DATA_FLUSH memref_t objects.  I'll clean it up.

brettcoon commented:
I fixed memref.h and cleaned up the helper routines in memref_gen.h, and then inexplicably the drcachesim config_reader_unit_test started failing on x86-32 due to a brittle floating-point comparison.  I don't think my changes could have _directly_ triggered this, but since equality tests on FP values are sensitive to compiler whims and other factors, I went ahead and included a fix for it in the PR.  I can split it out if needed, but my main PR can't pass regression without it.

derekbruening commented on clients/drcachesim/common/memref.h:
```diff
@@ -190,6 +190,13 @@ struct _memref_marker_t {
     uintptr_t marker_value; /**< A value whose meaning depends on the marker type. */
```
> The fix is to specify which union member is needed in the initialization statement.

The PR description (becomes the commit message) is now stale and needs updating.

brettcoon replied:
Done.

derekbruening commented on clients/drcachesim/common/memref.h:
```diff
@@ -190,6 +190,13 @@ struct _memref_marker_t {
 };
 
+/**
+ * To enable memref_t to be default initialized, a byte array with an initializer
```
Make a doxygen link to any types: `#memref_t`

derekbruening commented on clients/drcachesim/common/memref.h:
```diff
@@ -213,8 +220,12 @@ typedef union _memref_t {
     struct _memref_flush_t flush;      /**< A software-initiated cache flush. */
     struct _memref_thread_exit_t exit; /**< A thread exit. */
     struct _memref_marker_t marker;    /**< A marker holding metadata. */
+    uint8_t _raw_bytes[MEMREF_T_SIZE_BYTES] = {}; /**< Do not use: for init only. */
```
Could you clarify what is happening here according to the language specs: so at most one union variant can have a default member initializer, which presumably turns into the default constructor?  That means just saying "memref_t x;" calls the default constructor which zeroes _raw_bytes, right?

But now what about "memref_t x = {}": I assume there's copy elision and this will call a regular consructor: but which one?  The examples of unions have it initializing just the first union field.  Does it now, with the _raw_bytes having the default member initializer, call that constructor instead?  So it's now the same as "memref_t x;"?

Should we get rid of all the "= {};" below if we're not sure what's happening in that case and we're sure "memref_t x;' is now initialized the way we want?

Is this the way we want, having to zero all the bytes every time: what is the size of memref_instr_t?  Something like 70 bytes?  Does that turn into a memset call?  While memref_thread_exit_t is 24 bytes?  Are we now paying too much overhead?  Maybe it's fine.  Just trying to understand the tradeoffs.

derekbruening replied:
The later thread https://github.com/DynamoRIO/dynamorio/pull/7435#discussion_r2036568248 ended up discussing the same thing and showed that indeed with this default member initializer just 'memref_t x;` zeroes all 70-odd bytes.  That same thread also discusses what I brought up here about overhead.

brettcoon replied:
I assume that discussion subsumes this one.

brettcoon replied:
I will add that once C++20 is allowed, I believe the current approach in this PR enables an enthusiastic developer to optimize for their use of a smaller union member by initializing only that portion with a designated initializer like `memref_t memref = { .data = {} };`

brettcoon commented on clients/drcachesim/common/memref.h:
```diff
@@ -213,8 +220,12 @@ typedef union _memref_t {
     struct _memref_flush_t flush;      /**< A software-initiated cache flush. */
     struct _memref_thread_exit_t exit; /**< A thread exit. */
     struct _memref_marker_t marker;    /**< A marker holding metadata. */
+    uint8_t _raw_bytes[MEMREF_T_SIZE_BYTES] = {}; /**< Do not use: for init only. */
```
Regarding your first comment, the issue of padding bytes is why I put the default initializer on a new byte array rather than one of the member structs: a byte array won't have any padding bytes that might not get initialized.  I'll add some text to the comment to explain this.

Regarding your second comment, I believe the order of the entries is irrelevant in terms of overhead.  Furthermore, I'm pretty sure the initializer overhead is only paid when you ask for it, for example with a declaration like "memref_t memref = {};"

abhinav92003 replied:
> I'm pretty sure the initializer overhead is only paid when you ask for it, for example with a declaration like "memref_t memref = {};"

Isn't _raw_bytes always initialized due to the ` = {}`, even if there's no `= {}` on the corresponding memref_t object?

derekbruening replied:
This is what I claimed in https://github.com/DynamoRIO/dynamorio/pull/7435#discussion_r2035755195 as well.  I thought the member default initializer was added to all constructors (in this case the single default constructor) and would thus be executed for any declared object.

brettcoon replied:
I did some benchmarking and it appears you're correct: the member default initialization happens regardless of whether you specify `= {}` in your declaration or not if a union member has an initializer.

My test was a loop where each iteration declares a local memref_t and accesses one field in it.  I tried several variations w/ and w/o explicit initialization:

1. original memref_t
2. memref_t with `raw_bytes[] = {}` as *first* union member
3. memref_t with `raw_bytes[] = {}` as *last* union member
4. memref_t with `raw_bytes[]` as *last* union member
5. memref_t with `raw_bytes[]` as *first* union member

For each type I tried declaring the memref with and without initialization, e.g. `memref_t memref;` and `memref_t memref = {};`.

Bottom line, initialization costs about 4ns on a 4GHz AMD, no matter how it's done with initializers, or where the raw_bytes[] array is in the union member list.  FYI, using memset() to clear memref_t takes more like 11ns.

So, the question now is whether or not memref_t init safety is worth 4ns.  I don't have an answer.

derekbruening replied:
This is the same question brought up in https://github.com/DynamoRIO/dynamorio/pull/7435#discussion_r2035755195: whether the now-always-there overhead is worth it.  I think it would probably be fine to always initialize.  The analyzer serial loop does declare memref_t inside its loop, but we could move it if it mattered (probably not); the parallel declares it outside.  I also think it's fine to leave it uninitialized but have the `={}` properly zero a memref_t where the user wants to do it.

abhinav92003 replied:
> I think it would probably be fine to always initialize.

Xref https://github.com/DynamoRIO/dynamorio/pull/7114#discussion_r1886178344 where I wished for always-initialized fields in memref_t. So yes, it'd be nice to do that. But let's do it separately in a future PR.

> I also think it's fine to leave it uninitialized but have the ={} properly zero a memref_t where the user wants to do it.

I think that is indeed ensured if `_raw_bytes` is the first union member.

abhinav92003 commented on clients/drcachesim/common/memref.h:
```diff
@@ -203,18 +210,28 @@ struct _memref_marker_t {
-    struct _memref_thread_exit_t exit; /**< A thread exit. */
-    struct _memref_marker_t marker;    /**< A marker holding metadata. */
+    // The _raw_bytes entry is for initialization purposes and must be first in
+    // this list.  It is not intended to be used for memref_t access.
```
Also note the C nuance about padding bytes in the first union member not being initialized by a `= {}` which could actually hold an actual data member of some other union member, and hence that's why we need the raw bytes array.

brettcoon replied:
Done.

abhinav92003 commented on clients/drcachesim/common/memref.h:
```diff
@@ -203,18 +210,28 @@ struct _memref_marker_t {
 } memref_t;
 
+static_assert(sizeof(memref_t) == MEMREF_T_SIZE_BYTES,
+              "Update MEMREF_T_SIZE_BYTES to match sizeof(memref_t).");
```
To be more specific, also include: "Did the largest union member change?".

brettcoon replied:
Done.

abhinav92003 commented on clients/drcachesim/common/memref.h:
```diff
@@ -203,18 +210,28 @@ struct _memref_marker_t {
+ *
+ * Note that #memref_t is **not** initialized by default.  The _raw_bytes array
+ * is added to the union as its first member to make sure a #memref_t object
+ * can be easily initialized if desired, for example `memref_t memref = {};`
```
s/easily/fully/

derekbruening commented on clients/drcachesim/common/memref.h:
```diff
@@ -203,18 +210,31 @@ struct _memref_marker_t {
+ *
+ * Note that #memref_t is **not** initialized by default.  The _raw_bytes array
+ * is added to the union as its first member to make sure a #memref_t object
+ * can be fully initialized if desired, for example `memref_t memref = {};`
```
nit: Missing final `.`
